### PR TITLE
[2/5] Add support for specifying crate name for proc macro.

### PR DESCRIPTION
**Pull request by nlordell**
This adds new syntax to the `contract!` proc macro so you can specify a crate name for the runtime crate. This allows us to support rare cases where the crate is renamed either in `Cargo.toml` or `extern crate ethcontract as foo;`

This is needed to properly build doctests for our crate when we included the example contracts.

### CODE REVIEW

File: derive/src/lib.rs [hunk 1]
```diff
@@ -23,7 +32,43 @@ macro_rules! ident {
...
 }
 
-fn expand_contract(input: LitStr) -> Result<TokenStream> {
+struct ContractArgs {
+    artifact_path: LitStr,
+    runtime_crate: Option<Ident>,
+}
+
+impl Parse for ContractArgs {
+    fn parse(input: ParseStream) -> ParseResult<Self> {
```
Comment (fleupold) [inline derive/src/lib.rs:41]: Is it possible to test these functions maybe even doc tests? This would allow us to easier understand what's happening.

File: derive/src/lib.rs [hunk 2]
```diff
@@ -23,7 +32,43 @@ macro_rules! ident {
...
+        let mut result = ContractArgs {
+            artifact_path: input.parse()?,
+            runtime_crate: None,
+        };
+
+        while !input.is_empty() {
```
Comment (fleupold) [inline derive/src/lib.rs:47]: Do we really need a while loop here? Can there be multiple crate as definitions in one line and do we really want to take the last then?

File: derive/src/lib.rs [hunk 3]
```diff
@@ -23,7 +32,43 @@ macro_rules! ident {
...
+            input.parse::<Token![,]>()?;
+            if input.is_empty() {
+                break;
+            }
+
+            let param = input.call(Ident::parse_any)?;
+            input.parse::<Token![=]>()?;
```
Comment (fleupold) [inline derive/src/lib.rs:54]: What's the difference between `Token![=]` and `Token![,]`?

File: derive/src/lib.rs [hunk 4]
```diff
@@ -23,7 +32,43 @@ macro_rules! ident {
...
+    fn parse(input: ParseStream) -> ParseResult<Self> {
+        let mut result = ContractArgs {
+            artifact_path: input.parse()?,
+            runtime_crate: None,
+        };
+
+        while !input.is_empty() {
+            input.parse::<Token![,]>()?;
+            if input.is_empty() {
+                break;
```
Comment (fleupold) [inline derive/src/lib.rs:50]: Should we break here or also return an error?

File: derive/src/lib.rs [hunk 5]
```diff
@@ -33,11 +78,12 @@ fn expand_contract(input: LitStr) -> Result<TokenStream> {
...
     //   Eventually we can use the `Span::source_file` API to have a better
     //   experience.
-    let artifact_path = input.value();
-
-    let artifact = Artifact::load(&artifact_path)?;
+    let artifact_path = args.artifact_path;
+    let artifact = Artifact::load(&artifact_path.value())?;
     let contract_name = ident!(&artifact.contract_name.to_pascal_case());
 
+    let ethcontract = args.runtime_crate.unwrap_or_else(|| ident!("ethcontract"));
```
Comment (fleupold) [inline derive/src/lib.rs:85]: Is `crate` a better name for this?

File: derive/src/lib.rs [hunk 6]
```diff
@@ -255,30 +301,36 @@ fn expand_deploy(artifact: &Artifact) -> Result<TokenStream> {
...
-fn expand_function(artifact: &Artifact, function: &Function) -> Result<TokenStream> {
+fn expand_function(
+    artifact: &Artifact,
+    ethcontract: &Ident,
+    function: &Function,
+) -> Result<TokenStream> {
     let name = ident!(&function.name.to_snake_case());
     let name_str = Literal::string(&function.name);
 
+    let signature = function_signature(&function);
```
Comment (fleupold) [inline derive/src/lib.rs:312]: ubernit: this seems to be unrelated :wink:

File: common/src/truffle.rs [hunk 7]
```diff
@@ -81,7 +81,14 @@ pub struct Documentation {
     /// Contract documentation
     pub details: Option<String>,
     /// Contract method documentation.
-    pub methods: HashMap<String, String>,
+    pub methods: HashMap<String, DocEntry>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize)]
+/// A documentation entry.
+pub struct DocEntry {
```
Comment (fleupold) [inline common/src/truffle.rs:89]: Is this used?

**Review by fleupold — APPROVED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia1.giphy.com%2Fmedia%2FNHGZOqNepShCE%2Fgiphy.gif&token=awesome_sauce" data-gitmeme-token="awesome_sauce"></a>

Mostly macro related questions inline. I think adding some unit tests for the parsing functions could serve as great documentation.

File: derive/src/lib.rs (hunk 1 above)
Comment (nlordell) [reply derive/src/lib.rs:41]: Hmm, that is a good question. I can definitely look into it.

File: derive/src/lib.rs (hunk 2 above)
Comment (nlordell) [reply derive/src/lib.rs:47]: No we don't need a while loop. The reason I implemented it that way is we have open cases for adding code generation options and will need it eventually. See #10 for that issue.

File: derive/src/lib.rs (hunk 3 above)
Comment (nlordell) [reply derive/src/lib.rs:54]: `Token![=]` is the `=` token and `Token![,]` is the `,` token.

File: derive/src/lib.rs (hunk 4 above)
Comment (nlordell) [reply derive/src/lib.rs:50]: breaking here allows for trailing commas. So:
```rust
contract!(
    "path/to/artifact.json",
    foo = bar,
    baz = baw,
);
```
should work.

File: derive/src/lib.rs (hunk 5 above)
Comment (nlordell) [reply derive/src/lib.rs:85]: Maybe `runtime_crate`?

File: common/src/truffle.rs (hunk 7 above)
Comment (nlordell) [reply common/src/truffle.rs:89]: Yup!
https://github.com/gnosis/ethcontract-rs/blob/9fd4f359c110928774ad968620d0e27641f07651/derive/src/lib.rs#L313
